README.md
# WinCTS - Windows Configuration Tweak Scripts

Windows operating system tweaks is package of command line and powershell scripts.

## Functionality:
- Power Options - modify behavior of power buttons and power saving options.
- Computer Name Change - modify windows machine/computer name.
- AppxPackages - uninstall, permanently remove or reinstall Microsoft appx applications/packages.
- Settings - configuration, features and system behaviour "one-click" automated setup. Settings of this group are accessable and configurable from Windows User Interface (frontend) also. 
- Tweaks - configuration, features and system behaviour hidden settings not availabe from Windows User Interface. Organized into several configuration groups like privacy, user interface, updates, applications, hardware etc.

## Usage:
- Review and/or edit content of the config files in config subfolder
- Run particular cmd scripts with admin privileges

## AUTHOR:
Michal Selma <michal@selma.cc>


CHANGELOG.md
# CHANGELOG

## 0.4.1
* Appx configuration profiles added
* readme file fixed


## 0.4
* AppxPackages module added


## 0.3
* rebuild power options module
* split power options json config to configuration and dictionary files
* cosmetic changes to computer rename module


## 0.2.2
* logging added to computer change module
* cosmetic changes


## 0.2.1
* flat file configs replaced with json config type
* logging to file introduced
* minor improvements to computer rename module


## 0.2
* power options functionality
* computer name change functionality


## 0.1 - Initial Release
* project structure - files & folders 
* readme, license, info etc.



Tweaks.ps1
########################################
# Package: Windows Tweaks Script
# Type: CMD (Command Line) / PowerShell
# Platform: Windows 10 / Windows 11
# Source Code: https://github.com/michalselma/WinCTS
########################################

# Check admin rights and if needed relaunch script with admin privileges 
If (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]"Administrator")) {
	Write-Host "Script requires Administrator rights. Restarting..."
	Start-Process Tweaks.cmd -Verb RunAs
	Exit
}
Write-Host "Script is run with Administrator rights."

# Define config file location and name
$cfgfile = "$PSScriptRoot\config\tweaks.conf"

# Define log file location and name
$datetime = Get-Date -Format FileDateTime
$logfile = "$PSScriptRoot\log\$datetime-tweaks.log"

# Import script modules (functions)
$modules = Get-ChildItem -Recurse "$PSScriptRoot\tweaks\*.psm1" | Select-Object -ExpandProperty FullName
foreach ($module in $modules) {
	Import-Module -Name $module -ErrorAction Stop *>> $logfile
}

Write-Host "Processing tweaks configuration file..."
# Temporary variable to store config content
$cfgcontent = @()

# Read Config File
$cfgcontent = Get-Content $cfgfile -ErrorAction Stop

# Remove whitespaces at beggining and end of each object
$cfgcontent = $cfgcontent | ForEach-Object {$_.Trim()}

# Remove Empty lines/objects
$cfgcontent = $cfgcontent | Where-Object {$_}

# Remove comment lines from config file
$cfgcontent = $cfgcontent | Where-Object { $_.Substring(0,1) -ne '#'}

# Setting-up each option
foreach ($item in $cfgcontent) {
	# Split each object on '=' and build key & value pairs
	$item = $item.split("=")
	# Remove wihtespaces at beggining and end of each app after split
	$item = $item.trim()
	# Build key-value 'pairs' (not necessary - just to keep clear code)
	$key=$item[0]
	$value=$item[1]
	
	# Disable
	if ($value -eq 0) {
		$callfunction = "$key-Disable"
		Write-Host "Executing $callfunction"
		Invoke-Expression $callfunction *>> $logfile
	}
	# Enable
	elseif ($value -eq 1){
		$callfunction= "$key-Enable"
		Write-Host "Executing $callfunction"
		Invoke-Expression $callfunction *>> $logfile
	}
	# Skip
	elseif ($value -eq 2){
		Write-Host "Skipping $key modification."
		Write-Host "Skip -> $key" *>> $logfile
	}
	# Unknown
	else {
		Write-Host "*** $key - Incorrect or empty switch (='$value') in config file. Ignoring..."
		Write-Host "Error -> Unrecognized configuration switch for $key => '$value'" *>> $logfile
	}
}
# Restart explorer to apply changes
Write-Host "[INFO] Restarting explorer to apply changes" *>> $logfile
taskkill /f /im explorer.exe *>> $logfile
start explorer.exe *>> $logfile

Write-Host "[INFO] Script finished." *>> $logfile

# Ask to restart computer
$restart = Read-Host "Computer requires restart to appply changes. Input 'y' or 'yes' to restart computer now or ENTER to skip"
if ((('y', 'yes') -contains $restart)) {
	Write-Host "Restarting... "
	Restart-Computer
}
else {
	Write-Host "Skipping restart now. If any script changes were applied please restart your computer manually."
}


Settings.ps1
########################################
# Package: Windows Settings Script
# Type: CMD (Command Line) / PowerShell
# Platform: Windows 10
# Source Code: https://github.com/michalselma/WinCTS
########################################

# Check admin rights and if needed relaunch script with admin privileges 
If (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]"Administrator")) {
	Write-Host "Script requires Administrator rights. Restarting..."
	Start-Process Settings.cmd -Verb RunAs
	Exit
}
Write-Host "Script is run with Administrator rights."

# Define config file location and name
$cfgfile = "$PSScriptRoot\config\settings.conf"

# Define log file location and name
$datetime = Get-Date -Format FileDateTime
$logfile = "$PSScriptRoot\log\$datetime-settings.log"

# Import script modules (functions)
$modules = Get-ChildItem -Recurse "$PSScriptRoot\settings\*.psm1" | Select-Object -ExpandProperty FullName
foreach ($module in $modules) {
	Import-Module -Name $module -ErrorAction Stop *>> $logfile
}

Write-Host "Processing settings configuration file..."
# Temporary variable to store config content
$cfgcontent = @()

# Read Config File
$cfgcontent = Get-Content $cfgfile -ErrorAction Stop

# Remove whitespaces at beggining and end of each object
$cfgcontent = $cfgcontent | ForEach-Object {$_.Trim()}

# Remove Empty lines/objects
$cfgcontent = $cfgcontent | Where-Object {$_}

# Remove comment lines from config file
$cfgcontent = $cfgcontent | Where-Object { $_.Substring(0,1) -ne '#'}

# Setting-up each option
foreach ($item in $cfgcontent) {
	# Split each object on '=' and build key & value pairs
	$item = $item.split("=")
	# Remove wihtespaces at beggining and end of each app after split
	$item = $item.trim()
	# Build key-value 'pairs' (not necessary - just to keep clear code)
	$key=$item[0]
	$value=$item[1]
	
	# Disable
	if ($value -eq 0) {
		$callfunction = "$key-Disable"
		Write-Host "Executing $callfunction"
		Invoke-Expression $callfunction *>> $logfile
	}
	# Enable
	elseif ($value -eq 1){
		$callfunction= "$key-Enable"
		Write-Host "Executing $callfunction"
		Invoke-Expression $callfunction *>> $logfile
	}
	# Skip
	elseif ($value -eq 2){
		Write-Host "Skipping $key modification."
		Write-Host "Skip -> $key" *>> $logfile
	}
	# Unknown
	else {
		Write-Host "*** $key - Incorrect or empty switch (='$value') in config file. Ignoring..."
		Write-Host "Error -> Unrecognized configuration switch for $key => '$value'" *>> $logfile
	}
}
# Restart explorer to apply changes
Write-Host "Restarting explorer to apply changes"
taskkill /f /im explorer.exe *>> $logfile
start explorer.exe *>> $logfile

Write-Host "Script finished."

# Ask to restart computer
$restart = Read-Host "Computer requires restart to appply changes. Input 'y' or 'yes' to restart computer now or ENTER to skip"
if ((('y', 'yes') -contains $restart)) {
	Write-Host "Restarting... "
	Restart-Computer
}
else {
	Write-Host "Skipping restart now. If any script changes were applied please restart your computer manually."
}


AppxPackages.ps1
########################################
# Package: Windows Appx Packages Modification Script
# Type: CMD (Command Line) / PowerShell
# Platform: Windows 10 / Windows 11
# Source Code: https://github.com/michalselma/WinCTS
########################################

# Check admin rights and if needed relaunch script with admin privileges 
If (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]"Administrator")) {
	Write-Host "Script requires Administrator rights. Restarting..."
	Start-Process AppxPackages.cmd -Verb RunAs
	Exit
}
Write-Host "Script is run with Administrator rights."

# Define config file location and name
$cfgfile = "$PSScriptRoot\config\appxpackages-w10.conf"

# Define log file location and name
$datetime = Get-Date -Format FileDateTime
$logfile = "$PSScriptRoot\log\$datetime-appxpackages-w10.log"

# Import script modules (functions)
Import-Module -Name "$PSScriptRoot\AppxPackages.psm1" -ErrorAction Stop *>> $logfile

# Get list of installed applications/packages for reference/comaprision purpose
Write-Host "[INFO] Installed appx packages [before snapshot]:" *>> $logfile
Get-AppxPackage -AllUsers | Format-Table -Property Name *>> $logfile

Write-Host "Starting Applications installation/uninstallation module."

$removeapps, $addapps, $skipapps = ProcessAppxConfig($cfgfile)
Write-Host "Config file:" ($removeapps.length) "items for uninstallation."
Write-Host "Config file:" ($addapps.length) "items for installation/reinstallation."
Write-Host "Config file:" ($skipapps.length) "items to be skipped."

if (($removeapps.length) -eq 0){
	Write-Host "Nothing to be uninstalled."
}
else {
	Write-Host "Please choose uninstalltion/removal option:"
	Write-Host "1 - Current user uninstallation"
	Write-Host "2 - All existing users uninstallation"
	Write-Host "3 - Permanent Removal for all existing and future users"
	$removaloption = Read-Host "Input '1', '2' or '3' to continue or ENTER to cancel)"
	if ((('1') -contains $removaloption)) {
		UninstallAppx_currentuser($removeapps) *>> $logfile
	}
	elseif ((('2') -contains $removaloption)) {
		UninstallAppx_currentuser($removeapps) *>> $logfile
		UninstallAppx_allusers($removeapps) *>> $logfile
	}
	elseif ((('3') -contains $removaloption)) {
		UninstallAppx_currentuser($removeapps) *>> $logfile
		UninstallAppx_allusers($removeapps) *>> $logfile
		RemoveAppx_SysProvisioned($removeapps) *>> $logfile
	}
	else {
		Write-Host "Uninstall/Removal cancelled..." 
		Write-Host "[INFO] Uninstall/Removal cancelled..." *>> $logfile
	}
}

if (($addapps.length) -eq 0){
	Write-Host "Nothing awaiting installation."
	Write-Host "[INFO] Nothing awaiting installation." *>> $logfile
}
else {
	InstallAppx_currentuser($addapps) *>> $logfile
}

Write-Host "Finished Applications installation/uninstallation module."

# Get list of installed applications/packages for reference/comaprision purpose
Write-Host "[INFO] Installed appx packages [after snapshot]:" *>> $logfile
Get-AppxPackage -AllUsers | Format-Table -Property Name *>> $logfile

Write-Host "Script finished."

# Ask to restart computer
$restart = Read-Host "Computer requires restart to appply changes. Input 'y' or 'yes' to restart computer now or ENTER to skip"
if ((('y', 'yes') -contains $restart)) {
	Write-Host "Restarting... "
	Restart-Computer
}
else {
	Write-Host "Skipping restart now. If any script changes were applied please restart your computer manually."
}


PowerOptions.ps1
####################################################
# Package: Windows Power Options Script
# Type: CMD (Command Line) / PowerShell
# Platform: Windows 11
# Source Code: https://github.com/michalselma/WinCTS
# File Date: 2023-05-23
####################################################


### Check admin rights and if needed relaunch script with admin privileges 
If (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]"Administrator")) {
	Write-Host "[WARNING] Script requires Administrator rights. Restarting..."
	Start-Process PowerOptions.cmd -Verb RunAs
	Exit
}
Write-Host "[INFO] Script is run with Administrator rights."


# Define log file location and name
$logfile = "$PSScriptRoot\log\power.log"

$datetime = Get-Date -Format "dddd, yyyy-MM-dd HH:mm:ss.fff K"
Write-Output "[START] $datetime" *>> $logfile


# Get configuration file path
$presetpath = "$PSScriptRoot\config\power.json"

# Read configuration file
$json_config = Get-Content -Raw $presetpath -ErrorAction Stop | ConvertFrom-Json


# Get dictionary file path
$presetpath = "$PSScriptRoot\config\power-dict.json"

# Read dictionary file
$json_dict = Get-Content -Raw $presetpath -ErrorAction Stop | ConvertFrom-Json


# Get config choice from user
Write-Host "Choose power options preset type"
Write-Host "1 - Desktop"
Write-Host "2 - Laptop"
$configtype = Read-Host "Input preset value and press Enter"

if ((('1') -contains $configtype)) {
	$option = 'desktop'
}
elseif ((('2') -contains $configtype)) {
	$option = 'laptop'
}
else {
	Write-Output "[INFO] Incorrect option. Stopping script." | Tee-Object -FilePath $logfile -Append
	Exit
}


Write-Output "Preset type: $option" | Tee-Object -FilePath $logfile -Append


# Setting-up power options for chosen preset
foreach ($object in $json_config.PSObject.Properties) {
	$id = $object.Name
	
	# If chosen preset (desktop/laptop) is defined in option key in dictionary for object, start processing
	if (($json_dict.$id.option -contains $option)) {
		Write-Output "[INFO] Processing $id" | Tee-Object -FilePath $logfile -Append

		$type=$json_config.$id.type
		$key=$json_config.$id.key
		$value=$json_config.$id.value

		# Execute Windows 'powercfg.exe' with specific switch (type) and/or key-value data
		if ($type -eq "change") {
			Write-Output "Executing: powercfg /change $key $value" *>> $logfile
			powercfg /change $key $value
		}
		
		elseif ($type -eq "setdcvalueindex") {
			Write-Output "Executing: powercfg /setdcvalueindex SCHEME_CURRENT SUB_BUTTONS $key $value" *>> $logfile
			powercfg /setdcvalueindex SCHEME_CURRENT SUB_BUTTONS $key $value
		}
		
		elseif ($type -eq "setacvalueindex") {
			Write-Output "Executing: powercfg /setacvalueindex SCHEME_CURRENT SUB_BUTTONS $key $value" *>> $logfile
			powercfg /setacvalueindex SCHEME_CURRENT SUB_BUTTONS $key $value
		}
		
		elseif ($type -eq "other") {
			Write-Output "Executing: powercfg /$key $value" *>> $logfile
			powercfg /$key $value
		} 
		
		else {
			Write-Output "[ERROR] Incorrect configuration type $type for id $id" | Tee-Object -FilePath $logfile -Append
		}
	}

	else {
		Write-Output "Skipping $id" *>> $logfile
	}
}


$datetime = Get-Date -Format "dddd, yyyy-MM-dd HH:mm:ss.fff K"
Write-Output "[END] $datetime" *>> $logfile


ComputerNameChange.ps1
####################################################
# Package: Change Windows Computer Name Script
# Type: CMD (Command Line) / PowerShell
# Platform: Windows 11
# Source Code: https://github.com/michalselma/WinCTS
# File Date: 2023-04-29
####################################################


# Check admin rights and if needed relaunch script with admin privileges 
If (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]"Administrator")) {
	Write-Host "[WARNING] Script requires Administrator rights. Restarting..."
	Start-Process ComputerNameChange.cmd -Verb RunAs
	Exit
}
Write-Host "[INFO] Script is run with Administrator rights."


# Define log file location and name
$logfile = "$PSScriptRoot\log\computername.log"

$datetime = Get-Date -Format "dddd, yyyy-MM-dd HH:mm:ss.fff K"
Write-Output "[START] $datetime" *>> $logfile


# Get current computer name
$computerName = hostname
Write-Output "Current computer name: $computerName" | Tee-Object -FilePath $logfile -Append


# Start rename
$confirmRename = Read-Host "Do you want to change computer name ? (input 'y' or 'yes' and press ENTER to continue)"

if ((('y', 'yes') -contains $confirmRename)) {
	$newComputerName = Read-Host "Enter new computer name"
	Write-Output "Changing computer name to: $newComputerName" | Tee-Object -FilePath $logfile -Append
	Rename-Computer -NewName $newComputerName

	# Ask to restart computer
	$restart = Read-Host "Computer requires restart to appply changes. Input 'y' or 'yes' to restart computer now or ENTER to skip"
	if ((('y', 'yes') -contains $restart)) {
		Write-Output "Restarting..." | Tee-Object -FilePath $logfile -Append
		Restart-Computer
	}

	else {
		Write-Output "Skipping computer restart now. Please restart your computer later." | Tee-Object -FilePath $logfile -Append
	}
}

else {
	Write-Output "Undefined key pressed. Closing..." | Tee-Object -FilePath $logfile -Append
}


$datetime = Get-Date -Format "dddd, yyyy-MM-dd HH:mm:ss.fff K"
Write-Output "[END] $datetime" *>> $logfile

Appx.ps1
####################################################
# Package: Windows Appx Packages Modification Script
# Type: CMD (Command Line) / PowerShell
# Platform: Windows 11
# Source Code: https://github.com/michalselma/WinCTS
# File Date: 2023-05-23
####################################################


### Check admin rights and if needed relaunch script with admin privileges 
If (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]"Administrator")) {
	Write-Host "[WARNING] Script requires Administrator rights. Restarting..."
	Start-Process Appx.cmd -Verb RunAs
	Exit
}
Write-Host "[INFO] Script is run with Administrator rights." 


# Define log file location and name
$logfile = "$PSScriptRoot\log\appx.log"

$datetime = Get-Date -Format "dddd, yyyy-MM-dd HH:mm:ss.fff K"
Write-Output "[START] $datetime" *>> $logfile


# Get config choice from user
Write-Host "Choose removal profile:"
Write-Host "1 - Default (Home, but no or limited plugins, xbox, bing or ms apps)"
Write-Host "2 - Home"
Write-Host "3 - Work"
Write-Host "4 - Minimal"
$configtype = Read-Host "Input preset value and press Enter"

if ((('1') -contains $configtype)) {
	Write-Output "Processing preset type: Default" | Tee-Object -FilePath $logfile -Append
	# Configuration file path
	$presetpath = "$PSScriptRoot\config\appx.json"
}
elseif ((('2') -contains $configtype)) {
	Write-Output "Processing preset type: Home" | Tee-Object -FilePath $logfile -Append
	# Configuration file path
	$presetpath = "$PSScriptRoot\config\appx_home.json"
}
elseif ((('3') -contains $configtype)) {
	Write-Output "Processing preset type: Work" | Tee-Object -FilePath $logfile -Append
	# Configuration file path
	$presetpath = "$PSScriptRoot\config\appx_work.json"
}
elseif ((('4') -contains $configtype)) {
	Write-Output "Processing preset type: Minimal" | Tee-Object -FilePath $logfile -Append
	# Configuration file path
	$presetpath = "$PSScriptRoot\config\appx_min.json"
}
else {
	Write-Output "[INFO] Incorrect option. Stopping script." | Tee-Object -FilePath $logfile -Append
	Exit
}


# Read configuration file
$json_config = Get-Content -Raw $presetpath -ErrorAction Stop | ConvertFrom-Json


# Get dictionary file path
$presetpath = "$PSScriptRoot\config\appx-dict.json"

# Read dictionary file
# *** not in use yet
# $json_dict = Get-Content -Raw $presetpath -ErrorAction Stop | ConvertFrom-Json


# Read list of appx installed in OS
# $os_appx_list = Get-AppxPackage -AllUsers | Select-Object Name

# Write-Output "Found the following appx:" *>> $logfile
# Write-Output $os_appx_list *>> $logfile

# Read list of appx installed in OS. Distinct to get unique Names only
$os_appx_list_unique = Get-AppxPackage -AllUsers | Sort-Object -Property Name -Unique | Select-Object Name

Write-Output "Found the following unique appx:" *>> $logfile
Write-Output $os_appx_list_unique *>> $logfile


# Processing Get-AppxPackage -AllUsers
Write-Output "[INFO] Processing removal of AppxPackages for all users" | Tee-Object -FilePath $logfile -Append
foreach ($object in $os_appx_list_unique) {

	# Assign each appx name to $appx_name
	$appx_name=$object.Name

	Write-Output "Processing $appx_name" | Tee-Object -FilePath $logfile -Append

	# if configuration for $appx_name is specified in config file
	if ($json_config.$appx_name) {

		# if action is specified for $appx_name
		if ($json_config.$appx_name.action){

			# Read, check and if applicable action it
			$action=$json_config.$appx_name.action
			Write-Output "$appx_name is set: $action" *>> $logfile
			
			if ($action -eq "sys"){
				Write-Output "$appx_name is system application. Skipping..." *>> $logfile
			}
			
			elseif ($action -eq "skip"){
				Write-Output "$appx_name is set to skip. Skipping..." *>> $logfile
			}
			
			elseif ($action -eq "remove"){
				Write-Output "$appx_name is set to remove. Removing..." *>> $logfile
				$pkg_full_name = (Get-AppxPackage -AllUsers | Where-Object{$_.Name -eq $appx_name} | Select-Object PackageFullName).PackageFullName
				
				# Check if there is single or multiple FullPackage Names for appx (e.g. separate for x32 and x64)
				$count_pkg_full_name = @($pkg_full_name).Count
				Write-Output "There is/are $count_pkg_full_name FullPachageName(s) for $appx_name" *>> $logfile

				if ($count_pkg_full_name -eq 0){
					Write-Output "[ERROR] No FullPackages found for $appx_name" | Tee-Object -FilePath $logfile -Append
				}
				
				elseif ($count_pkg_full_name -eq 1){
					Write-Output "Package Name for $appx_name is $pkg_full_name" *>> $logfile
					Write-Output "[INFO] Removing $pkg_full_name" | Tee-Object -FilePath $logfile -Append
					Remove-AppxPackage -AllUsers -Package $pkg_full_name
				}
				
				else {
					$count = 0
					foreach ($name in $pkg_full_name) {
						$count = $count + 1
						Write-Output "Package Name #$count for $appx_name is $name" *>> $logfile
						Write-Output "[INFO] Removing $name" | Tee-Object -FilePath $logfile -Append
						Remove-AppxPackage -AllUsers -Package $name
					}
				}
			}
			
			else {Write-Output "[ERROR] Unknown action type $action for $appx_name" | Tee-Object -FilePath $logfile -Append}
		}
		
		else {Write-Output "[WARNING] No action found in configuration file for $appx_name" | Tee-Object -FilePath $logfile -Append}
	}
	
	else {Write-Output "[WARNING] No configuration for $appx_name" | Tee-Object -FilePath $logfile -Append}
}

# Read list of provisioned appx installed in OS.
# $os_appx_prov_list = Get-AppxProvisionedPackage -Online | Select-Object DisplayName

# Write-Output "Found the following provisioned appx:" *>> $logfile
# Write-Output $os_appx_prov_list *>> $logfile


# Read list of provisioned appx installed in OS. Distinct to get unique DisplayNames only
$os_appx_prov_list_unique = Get-AppxProvisionedPackage -Online | Sort-Object -Property DisplayName -Unique | Select-Object DisplayName

Write-Output "Found the following unique provisioned appx:" *>> $logfile
Write-Output $os_appx_prov_list_unique *>> $logfile


# Processing Get-AppxProvisionedPackage -Online
Write-Output "[INFO] Processing removal of AppxProvisionedPackages" | Tee-Object -FilePath $logfile -Append
foreach ($object in $os_appx_prov_list_unique) {

	# Assign each appx name to $appx_name
	$appx_prov_name=$object.DisplayName

	Write-Output "Processing $appx_prov_name" | Tee-Object -FilePath $logfile -Append

	# if configuration for $appx_name is specified in config file
	if ($json_config.$appx_prov_name) {

		# if action is specified for $appx_name
		if ($json_config.$appx_prov_name.action) {

			# Read, check and if applicable action it
			$action=$json_config.$appx_prov_name.action
			Write-Output "$appx_prov_name is set: $action" *>> $logfile
			
			if ($action -eq "sys"){
				Write-Output "$appx_prov_name is system application. Skipping..." *>> $logfile
			}
			
			elseif ($action -eq "skip"){
				Write-Output "$appx_prov_name is set to skip. Skipping..." *>> $logfile
			}
			
			elseif ($action -eq "remove"){
				Write-Output "$appx_prov_name is set to remove. Removing..." *>> $logfile
				$prov_pkg_full_name = (Get-AppxProvisionedPackage -Online | Where-Object{$_.DisplayName -eq $appx_prov_name} | Select-Object PackageName).PackageName
				Write-Output "Package Name for $appx_prov_name is $pkg_full_name" *>> $logfile
				Write-Output "[INFO] Removing $prov_pkg_full_name" | Tee-Object -FilePath $logfile -Append
				Remove-AppxPackage -AllUsers -Package $prov_pkg_full_name
			}
			
			else {Write-Output "[ERROR] Unknown action type $action for $appx_prov_name" | Tee-Object -FilePath $logfile -Append}
		}
		
		else {Write-Output "[WARNING] No action found in configuration file for $appx_prov_name" | Tee-Object -FilePath $logfile -Append}
	}
	
	else {Write-Output "[WARNING] No configuration for $appx_prov_name" | Tee-Object -FilePath $logfile -Append}
}


$datetime = Get-Date -Format "dddd, yyyy-MM-dd HH:mm:ss.fff K"
Write-Output "[END] $datetime" *>> $logfile
